File: 3-Sum/3sum-sol.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    public List<List<Integer>> threeSum(int[] A) {
        // Sorting the initial array 
        Arrays.sort(A); 
        // Creating a hashset to hold seen triplets 
        Set<List<Integer>> allThreeSums = new HashSet<>(); 
        int secondLastIndex = A.length - 2;

        for(int i = 0; i < secondLastIndex; i++) {
            findTwoSum(i, A, allThreeSums); 
        }

        return new ArrayList<>(allThreeSums); 
    }

    private void findTwoSum(int rootIndex, int[] A, Set<List<Integer>> allThreeSums) {
        int left = rootIndex + 1; 
        int right = A.length - 1; 

        while (left < right) {
            int threeNumberSum = A[rootIndex] + A[right] + A[left]; 

            if(threeNumberSum == 0) {
                allThreeSums.add(Arrays.asList(A[rootIndex], A[left], A[right]));
                left++; 
                right--;

            } else if (threeNumberSum < 0) {
                left++;
            } else {
                right--; 
            }
        }
    }
}

File: 3-Sum/3sum-sol.py
from functools import reduce 

class Solution: 
    def threeSum(self, A): 

        # Sort the input array 
        A.sort()

        # Track the unique sets and seen 
        seen = set()
        all_three_sums = []
        second_last_index = len(A) - 2

        for i in range(0, second_last_index):
            self.find_two_sum(i, A, all_three_sums, seen)

        return all_three_sums


    def find_two_sum(self, root_index, A, all_three_sums, seen): 
        # Establish left and right pointer 
        left = root_index + 1
        right = len(A) - 1 

        # Search through the array until we have exhausted
        while left < right: 
            three_number_sum = A[root_index] + A[left] + A[right]

            if(three_number_sum == 0):
                number_list = [A[root_index], A[left], A[right]]
                number_list.sort()

                # Creating a string signature, ie '-121' to use in set 
                signature = reduce(lambda acc, num: str(acc) + 
                    str(num), number_list)
                
                if signature not in seen: 
                    all_three_sums.append(number_list)
                    seen.add(signature)

            # Moving the pointers 
                left += 1
                right -= 1
            elif three_number_sum < 0:
                left += 1
            else:
                right -= 1